### ISSUE TITLE
core: add support for CSS shorthands

### ISSUE THREAD

**Issue by layershifter** [microsoft/griffel#531]
Currently, Griffel does not support CSS shorthands, https://griffel.js.org/react/guides/limitations#css-shorthands-are-not-supported 😿  The plan is to bring the support back 🚀 

## How it works?

_This concept was inspired by StyleX and Fela ([fela-enforce-longhands](https://www.npmjs.com/package/) plugin)._

There are two scenarios where CSS longhands & shorthands could conflict, and the strategies for both are described below ⬇️

### Longhand should override

As we depend on JavaScript order of definitions for overrides, we should adhere to it:

```js
const useClasses = makeStyles({
  root: {
    padding: 0,
    paddingLeft: '10px',
  },
});
```

In this case, we will generate the following CSS and apply both classes to the element:

```html
<style data-priority="-1">
  .padding-0 {
    padding: 0;
  }
</style>
<style data-priority="0">
  .padding-left-10px {
    padding-left: 10px;
  }
</style>

<div class="padding-0 padding-left-10px"></div>
```

> Note: The order of CSS rules ensures that `padding-left` will override padding, resulting in a deterministic outcome.

### Shorthand should override

```js
const useClasses = makeStyles({
  root: {
    paddingLeft: '10px',
    padding: 0,
  },
});
```

Unlike the previous example, we will generate the following CSS and apply only a single class to the element:

```html
<style data-priority="-1">
  .padding-0 {
    padding: 0;
  }
</style>

<div class="padding-0"></div>
```

This occurs because `padding` overrides all its shorthands (`padding-left`, `padding-right`, etc.), eliminating the need to apply them individually.

### Order adjustments

All CSS shorthands will be defined with a lower priority (`-1` or lower) than their corresponding longhands (`0`). Rules with lower priority appear first in the CSS file (or DOM) and will be overridden by rules with higher priority.

> Note: Negative priorities are chosen for backwards compat. 

_This change has been implemented in #539._

### Shorthands & "longhands reset"

_Support for rules reset was added in #450._

Whenever a shorthand is defined, its corresponding longhands will be reset (refer to "Shorthand should override"):

```js
const useClasses = makeStyles({
  root: {
    padding: 0,

    // "padding" automatically resets the following longhands
    paddingTop: RESET,
    paddingRight: RESET,
    paddingBottom: RESET,
    paddingLeft: RESET,
  },
});
```

### Limitations on `border-*` shorthands

Unlike other CSS properties, `border-*` shorthands has an "equality" behavior, for instance, `border-left` & `border-color`:

```js
const style = {
  borderLeft: '1px solid red',
  // is essentially ⬇️
  borderLeftWidth: '1px',
  borderLeftStyle: 'solid',
  borderLeftColor: 'red',
};
```

```js
const style = {
  borderColor: 'red',
  // is essentially ⬇️
  borderTopColor: 'red',
  borderRightColor: 'red',
  borderBottomColor: 'red',
  borderLeftColor: 'red',
};
```

This behavior poses an issue as `borderLeft` cannot reset `borderColor` since it's a subset of `border*Color`, resulting in collision once these rules are applied:

```js
// How to handle this? 🫤
const style = {
  borderLeft: '1px solid green',
  borderColor: 'red',
}
```

### Rejected approach

One of the options considered was to define `borderColor`, `borderStyle` & `borderWidth` with lower priority than `borderLeft` & etc. to ensure deterministic behavior. However, this approach would be confusing, as illustrated in the example below:

```html
<style data-priority="-2">
  .border-color-red {
    border-color: red;
  }
</style>
<style data-priority="-1">
  .border-left-1px-solid-green {
    border-left: 1px solid green;
  }
</style>
```

Because of CSS order, both sets of rules would visually produce the same result:

```js
const useClasses = makeStyles({
  rootA: {
    borderLeft: '1px solid green',
    borderColor: 'red',
  },
  rootB: {
    borderColor: 'red',
    borderLeft: '1px solid green',
  },
});
```

While the outcome would be deterministic, it would be confusing for developers, as they would expect `borderColor` to override `borderLeft` in the first case, as is the case with other properties.

### Solution

Instead, we opted to continue maintain `borderColor`, `borderStyle` & `borderWidth` as unsupported shorthands i.e. they will retain their current limitations. Directional border shorthands will be fully supported (`borderLeft`, `borderTop`, etc.), ensuring that styles work as expected:

```js
const useClasses = makeStyles({
  rootA: {
    borderLeft: '1px solid green',
    ...shorthands.borderColor('red'),
  },
  rootB: {
    ...shorthands.borderColor('red'),
    borderLeft: '1px solid green',
  },
});
```

**Comment by ivi-hamiti** [comment]
I noticed the codemod is not published in npm, is there any support for automatic migration or has this to be done manually?

**Comment by ivi-hamiti** [comment]
Noticed there is an issue with the following shorthands codemod migration, if the last parameter is 0 it is omitted

```
{
  ...shorthands.gap('4px', 0),
  ...shorthands.margin('4px', 0),
  ...shorthands.padding('4px', 0)
}
```

is migrated to

```
{
  gap: '4px',  // wrong should had been '4px 0'
  margin: '4px', // wrong should had been '4px 0'
  padding: '4px' // wrong should had been '4px 0'
}
```

Second problem is that `shorthands.gap` parameter order is wrongly migrated

```
{
  ...shorthands.gap('4px', '8px')
}
```

is migrated to

```
{
  gap: '4px 8px' // wrong should had been '8px 4px'
}
```

based on [mdn gap docs](https://developer.mozilla.org/en-US/docs/Web/CSS/gap#formal_syntax) the syntax of gap is

```
gap: <row-gap> <column-gap>
```

while the [gap function definition](https://github.com/microsoft/griffel/blob/main/packages/core/src/shorthands/gap.ts) has it reverse

**Comment by layershifter** [comment]
> Noticed there is an issue with the following shorthands codemod migration, if the last parameter is 0 it is omitted

I cannot reproduce this using `griffel-codemod-shorthands@1.0.15`. Can you please check using that version?


> Second problem is that `shorthands.gap` parameter order is wrongly migrated

Ouch, indeed. Pushed a fix and released `griffel-codemod-shorthands@1.0.16` 👍

**Comment by ivi-hamiti** [comment]
was able to reproduce with `griffel-codemod-shorthands@1.0.16` but instead of string values i had tokens values

```
import { makeStyles, shorthands, tokens } from '@fluentui/react-components';

export const useStyles = makeStyles({
    root: {
        ...shorthands.margin(tokens.spacingVerticalM, 0),
    },
});
```

this is wrongly migrated to 

```
import { makeStyles, tokens } from '@fluentui/react-components';

export const useStyles = makeStyles({
    root: {
        margin: tokens.spacingVerticalM,
    },
});
```

**Comment by layershifter** [comment]
@ivi-hamiti thanks, fixed in `griffel-codemod-shorthands@1.0.17` 👍
